modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointLocal.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.transactions;
+
+import java.util.LinkedHashMap;
+import org.apache.ignite.internal.util.tostring.GridToStringInclude;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Transaction savepoint implementation for near local transactions.
+ */
+public class TxSavepointLocal implements TxSavepoint {
+    /** Savepoint ID */
+    @GridToStringInclude
+    private final String name;
+
+    /** Per-transaction read map. */
+    @GridToStringInclude
+    private final Map<IgniteTxKey, IgniteTxEntry> txMapSnapshot;
+
+    /**
+     * @param name Savepoint ID.
+     * @param tx Transaction, which state should be saved.
+     */
+    public TxSavepointLocal(String name, IgniteInternalTx tx) {
+        this.name = name;
+
+        Collection<IgniteTxEntry> stateEntries = tx.txState().allEntries();
+
+        txMapSnapshot = new LinkedHashMap<>(stateEntries.size());
+
+        putCopies(stateEntries, txMapSnapshot);
```

Comment:
consider u have a couple savepoints. Second savepoint contain all entries from first savepoint. It seems to be duplication and we can optimize tx entries storage: for instance, store certain entries and have a link to previous savepoint.